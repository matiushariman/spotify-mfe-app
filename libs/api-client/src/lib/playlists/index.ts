export * from './playlists';
export * from './playlists.types';
