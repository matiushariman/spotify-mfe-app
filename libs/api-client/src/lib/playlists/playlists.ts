import axios, { AxiosInstance, AxiosError, isAxiosError } from 'axios';

import { API_BASE_URL } from '../../constants';

import type { GetCurrentUserPlaylistsResponse } from './playlists.types';
import type { ApiClassConstructor } from '../../types/ApiClassConstructor.types';

export class PlaylistsApi {
  private readonly http: AxiosInstance;

  constructor({ token }: ApiClassConstructor) {
    this.http = axios.create({
      baseURL: `${API_BASE_URL}/v1`,
      headers: {
        'Content-Type': 'application /json',
        Authorization: `Bearer ${token}`,
      },
    });
  }

  /**
   * Get a list of the playlists owned or followed by the current Spotify user.
   *
   * - {@link https://developer.spotify.com/documentation/web-api/reference/#/operations/get-a-list-of-current-users-playlists|API Reference - Get Current User Playlists}
   */
  async getCurrentUserPlaylists() {
    return this.http
      .get<GetCurrentUserPlaylistsResponse>('/me/playlists')
      .then((res) => res)
      .catch((err: Error | AxiosError) => {
        if (isAxiosError(err)) {
          return Promise.reject(err?.response?.data);
        }

        return Promise.reject(err);
      });
  }
}
