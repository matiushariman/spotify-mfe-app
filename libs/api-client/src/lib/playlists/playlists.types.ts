import type { Image } from '../../types/Images.type';
import type { ExternalUrls } from '../../types/ExternalUrls.types';
import type { Followers } from '../../types/Followers.types';
import type { ApiErrorResponse } from '../../types/ApiErrorResponse.types';

interface PlaylistItem {
  /**
   * - The name of the playlist.
   */
  readonly name: string;
  /**
   * - true if the owner allows other users to modify the playlist.
   */
  readonly collaborative: boolean;
  /**
   * - The playlist description. Only returned for modified, verified playlists, otherwise null.
   */
  readonly description: string | null;
  /**
   * - Known external URLs for this playlist.
   */
  readonly external_urls: ExternalUrls;
  /**
   * - A link to the Web API endpoint providing full details of the playlist.
   */
  readonly href: string;
  /**
   * - The Spotify ID for the playlist.
   */
  readonly id: string;
  /**
   * - The user who owns the playlist
   */
  readonly owner: {
    /**
     * - Known public external URLs for this user.
     */
    readonly external_urls: ExternalUrls;
    /**
     * - Information about the followers of the users.
     */
    readonly followers: Followers;
    /**
     * - A link to the Web API endpoint for this user.
     */
    readonly href: string;
    /**
     * - The Spotify user ID for this user.
     */
    readonly id: string;
    /**
     * - The object type. Allowed value: "user"
     */
    readonly type: 'user';
    /**
     * - The Spotify URI for this user.
     */
    readonly uri: string;
    /**
     * - The name displayed on the user's profile. null if not available.
     */
    readonly display_name: string | null;
  };
  /**
   * The playlist's public/private status:
   * - `True` the playlist is public
   * - `False` the playlist is private
   * - `null` the playlist status is not relevant
   */
  readonly public: boolean | null;
  /**
   * - The version identifier for the current playlist. Can be supplied in other requests to target a specific playlist version
   */
  readonly snapshot_id: string;
  /**
   * - A collection containing a link ( href ) to the Web API endpoint where full details of the playlist's tracks can be retrieved, along with the total number of tracks in the playlist.
   * - Note, a track object may be `null`. This can happen if a track is no longer available.
   */
  readonly tracks: {
    /**
     * - A link to the Web API endpoint where full details of the playlist's tracks can be retrieved.
     */
    readonly href: string;
    /**
     * - Number of tracks in the playlist.
     */
    readonly total: number;
  } | null;
  /**
   * - The object type: "playlist"
   */
  readonly type: 'playlist';
  /**
   * - The Spotify URI for the playlist.
   */
  readonly uri: string;
  /**
   * - Images for the playlist. The array may be empty or contain up to three images. The images are returned by size in descending order.
   */
  readonly images: Image[];
}

export interface GetCurrentUserPlaylistsResponse {
  /**
   * - The requested content
   */
  readonly items: PlaylistItem[];
  /**
   * - A link to the Web API endpoint returning the full result of the request
   */
  readonly href: string;
  /**
   * - The maximum number of items in the response (as set in the query or by default).
   */
  readonly limit: number;
  /**
   * - URL to the next page of items. ( null if none)
   */
  readonly next: string | null;
  /**
   * - The offset of the items returned (as set in the query or by default)
   */
  readonly offset: number;
  /**
   * - URL to the previous page of items. ( null if none)
   */
  readonly previous: string;
  /**
   * - The total number of items available to return.
   */
  readonly total: number;
}

export type GetCurrentUserPlaylistsError = ApiErrorResponse;
