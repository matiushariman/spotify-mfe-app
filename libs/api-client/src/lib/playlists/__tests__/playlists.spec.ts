import { PlaylistsApi } from '../playlists';

describe('lib/playlists', () => {
  const playlistsApi = new PlaylistsApi({ token: '' });

  describe('getCurrentUserPlaylists()', () => {
    it('should return user playlists data', async () => {
      const res = await playlistsApi.getCurrentUserPlaylists();

      expect(res.status).toEqual(200);
      expect(res.data).toMatchSnapshot();
    });
  });
});
