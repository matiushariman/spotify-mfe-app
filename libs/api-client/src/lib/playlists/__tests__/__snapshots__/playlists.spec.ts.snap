// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/playlists getCurrentUserPlaylists() should return user playlists data 1`] = `
Object {
  "href": "https://api.spotify.com/v1/me/shows?offset=0&limit=20
",
  "items": Array [
    Object {
      "collaborative": true,
      "description": "string",
      "external_urls": Object {
        "spotify": "string",
      },
      "href": "string",
      "id": "string",
      "images": Array [
        Object {
          "height": 300,
          "url": "https://i.scdn.co/image/ab67616d00001e02ff9ca10b55ce82ae553c8228",
          "width": 300,
        },
      ],
      "name": "string",
      "owner": Object {
        "display_name": "string",
        "external_urls": Object {
          "spotify": "string",
        },
        "followers": Object {
          "href": "string",
          "total": 0,
        },
        "href": "string",
        "id": "string",
        "type": "user",
        "uri": "string",
      },
      "public": true,
      "snapshot_id": "string",
      "tracks": Object {
        "href": "string",
        "total": 0,
      },
      "type": "playlist",
      "uri": "string",
    },
  ],
  "limit": 20,
  "next": "https://api.spotify.com/v1/me/shows?offset=1&limit=1",
  "offset": 0,
  "previous": "https://api.spotify.com/v1/me/shows?offset=1&limit=1",
  "total": 4,
}
`;
